### PR TITLE
Replace deprecated Flask-oauthlib with authlib (Fixes #1100)

### LINKED ISSUE #1100: flask-oauthlib is deprecated in favour of authlib
The author of `flask-oauthlib` has deprecated it in favour of `Authlib`. <https://github.com/lepture/authlib>

There is  [a migration guide](https://blog.authlib.org/2018/migrate-flask-oauthlib-client-to-authlib).

`Authlib` doesn't currently have a compatible license, but from v0.11 will be released under a BSD license, which would be  compatible.

I suspect that this issue will need to be resolved before we can migrate to later versions of `oauthlib`.

### PR DESCRIPTION
**Pull request by rbubley**
I’ve only tested this with Google log-in. It would be good for someone to test other oauth providers before merging.

### CODE REVIEW

**Comment by lipis** [comment]
WOW!! Thanks man for this.. I'll test it with other providers..!

**Comment by lipis** [comment]
https://gae-init.appspot.com/signin/

Some of the providers are not working currently.. :(

**Comment by lipis** [comment]
(I haven't deployed your fix)

**Comment by lipis** [comment]
I tried locally with GitHub, Dropbox and getting the same error on click:

```
AttributeError
AttributeError: 'module' object has no attribute 'OAuthException'

Traceback (most recent call last)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 2449, in wsgi_app
response = self.handle_exception(e)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask_restful/__init__.py", line 269, in error_router
return original_handler(e)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 1866, in handle_exception
reraise(exc_type, exc_value, tb)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 2446, in wsgi_app
response = self.full_dispatch_request()
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 1951, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask_restful/__init__.py", line 269, in error_router
return original_handler(e)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 1820, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 1949, in full_dispatch_request
rv = self.dispatch_request()
File "/Users/lipis/work/lipis/gae-init/init/main/lib/flask/app.py", line 1935, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "/Users/lipis/work/lipis/gae-init/init/main/auth/github.py", line 41, in signin_github
return auth.signin_oauth(github)
File "/Users/lipis/work/lipis/gae-init/init/main/auth/auth.py", line 353, in signin_oauth
except oauth.OAuthException:
AttributeError: 'module' object has no attribute 'OAuthException'
```

**Comment by lipis** [comment]
Still getting an error.. try to setup GitHub with your local machine if you want to test it.

**Comment by rbubley** [comment]
Getting further but still not working yet...

**Comment by rbubley** [comment]
I suspect the remaining issue is due to `authlib` using `requests` and  `requests` requiring monkeypatching to work properly on GAE ([GAE docs on this issue](https://cloud.google.com/appengine/docs/standard/python/issue-requests#issuing_an_http_request))

**Comment by lipis** [comment]
That's kind of pain in the ass..!

**Comment by lipis** [comment]
Can you submit an issue there asking what the solution for GAE?

**Comment by lipis** [comment]
not comfortable monkey patching the app for it..

**Comment by rbubley** [comment]
With regard to the monkeypatching:

1. Monkeypatching is google's preferred solution (https://cloud.google.com/appengine/docs/standard/python/issue-requests#issuing_an_http_request)
2. It is one of two approaches recommended by `requests` (see https://toolbelt.readthedocs.io/en/latest/adapters.html#appengineadapter)

If you don't like the monkeypatch I can certainly try their alternative!

**Comment by lipis** [comment]
Because we are using the build in url fetch I would like to avoid it if possible.. is there an alternative?!

(In the meantime I deleted a few.. so you can update with master)

**Comment by rbubley** [comment]
I don't think there is an alternative (other than to stay with the deprecated Flask-oauthlib) -- but this is really just a switch from `requests-oauthlib` to `requests` + `requests_toolbelt`, so not really any less pure an approach. (Behind the scenes, `requests_toolbelt` switches it to using native urlfetch, I believe.)

**Comment by lipis** [comment]
Let's do it.. as long as everything else is working.. for all these providers...

**Comment by rbubley** [comment]
I've only checked github and google, as I haven't had a chance to get credentials for all the others. I have visually cross-checked what we have with [authlib/loginpass](https://github.com/authlib/loginpass/), and I don't see any reason for it not to work elsewhere.

File: main/auth/bitbucket.py [hunk 1]
```diff
@@ -14,33 +14,28 @@
...
 def bitbucket_authorized():
-  response = bitbucket.authorized_response()
-  if response is None:
+  id_token = bitbucket.authorize_access_token()
+  if id_token is None:
     flask.flash('You denied the request to sign in.')
     return flask.redirect(util.get_next_url())
 
-  flask.session['oauth_token'] = (response['access_token'], '')
+  flask.session['oauth_token'] = (id_token, '')
```
Comment (lepture) [inline main/auth/bitbucket.py]: you don't need this code.

File: main/auth/facebook.py [hunk 2]
```diff
@@ -25,22 +25,17 @@
...
 def facebook_authorized():
-  response = facebook.authorized_response()
-  if response is None:
+  id_token = facebook.authorize_access_token()
+  if id_token is None:
     flask.flash('You denied the request to sign in.')
     return flask.redirect(util.get_next_url())
 
-  flask.session['oauth_token'] = (response['access_token'], '')
+  flask.session['oauth_token'] = (id_token, '')
```
Comment (lepture) [inline main/auth/facebook.py]: you don't need this code.

**Comment by lipis** [comment]
https://gae-init.appspot.com/signin/

Only GAE, Bitbucket and GitHub are working.. I deployed the latest.. they were working before that.

File: main/auth/bitbucket.py (hunk 1 above)
Comment (lipis) [inline main/auth/bitbucket.py]:
```suggestion
```

File: main/auth/facebook.py (hunk 2 above)
Comment (lipis) [inline main/auth/facebook.py]:
```suggestion
```

**Comment by rbubley** [comment]
I see. For OAuth1 code currently does not store tokens. The way this is done currently in master is to store them in Flask.session, but this is apparently insecure. Will ponder...

**Comment by lipis** [comment]
where should we store them? Is there a light in the tunnel?

**Comment by rbubley** [comment]
I think it will be best if in this PR we stick with session storage, and just get `authlib` working. Then I will open a separate PR in a while to improve the security.

**Comment by rbubley** [comment]
Previously @lipis had this branch working with GAE, Bitbucket and GitHub; with the latest changes I have also tested and had it working with twitter.

**Comment by lipis** [comment]
will treat tomorrow! one provider at the time

**Comment by lepture** [comment]
@rbubley according to your usage, you don't have to store any tokens for OAuth 2. But you need to store request token for OAuth 1.

**Comment by rbubley** [comment]
Thanks for looking @lepture. That was indeed my intention; hopefully the code reflects that.

File: main/auth/github.py [hunk 3]
```diff
@@ -56,3 +50,15 @@ def retrieve_user_from_github(response):
...
+#localhost:3000 vs 127.0.0.1:8080 issue
+# http://127.0.0.1:8080/api/auth/callback/github/
+
+# Need to compare what gets returned by 
+# flask.url_for(
+#      '%s_authorized' % oauth_app.name, _external=True, _scheme=scheme
+#    )
+
+# With the URL on the auth config page
+#http://localhost:3000/api/auth/callback/github/?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-authorization-request-errors%2F%23redirect-uri-mismatch&state=r1awNd3oT3XgWMLIX1ZheZHenIEvJZ
```
Comment (lipis) [inline main/auth/github.py]: this comment should go.. test the oauth on 127.0.0.1:8080 or localhost:8080

**Comment by lipis** [comment]
https://gae-init.appspot.com/signin/ Uploaded the latest.. Twitter works now, but not Facebook, Instagram and Microsoft

**Comment by lipis** [comment]
Facebook works!

**Comment by lipis** [comment]
Instagram still fails.. deployed the latest: https://gae-init.appspot.com/signin/

**Comment by rbubley** [comment]
I can't debug instagram, as

(1)  using your deployed version, I just get a sandbox error (https://gae-init-testing.appspot.com) {"error_type": "OAuthForbiddenException", "code": 403, "error_message": "You are not a sandbox user of this client"} 

(2) with the instagram API deprecated (ceasing to work in "early 2020"), I can't sign up for a new app.

Maybe PM/slack me the error message you get?

**Comment by lipis** [comment]
Maybe we can remove the instgram login.. it's facebook anyway!

**Comment by lipis** [comment]
https://gae-init.appspot.com/signin/ The latest.. Microsoft almost worked..! :)

**Comment by lipis** [comment]
https://github.com/gae-init/gae-init/pull/1224

**Comment by rbubley** [comment]
For microsoft, I updated from the deprecated :"live.com" endpoints to "microsoftonline.com". From the error message, it looks like you may need to re-register the app with microsoft. Not an ideal upgrade path, I'm afraid.

**Comment by lipis** [comment]
Sync with master to deprecate Inatagram

**Comment by lipis** [comment]
Microsoft works!
https://gae-init.appspot.com/signin/

**Review by lipis — APPROVED** [review]

**Comment by lipis** [comment]
So this is ready?

File: main/auth/auth.py [hunk 4]
```diff
@@ -312,11 +313,11 @@ def urls_for_oauth(next_url):
...
-  upper_name = name.upper()
-  app.config[upper_name] = service_config
-  service_oauth = oauth.OAuth()
-  service_app = service_oauth.remote_app(name, app_key=upper_name)
-  service_oauth.init_app(app)
+  #upper_name = name.upper()
+  #app.config[upper_name] = service_config
+  service_oauth = oauth.OAuth(app)
+  service_app = service_oauth.register(name, **service_config)
+  #service_oauth.init_app(app)
```
Comment (lipis) [inline main/auth/auth.py]: remove commented out code.

**Comment by lipis** [comment]
@rbubley merginnnn?

**Pull request merged**